bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java

```diff
@@ -90,137 +85,82 @@ protected void deactivate(ComponentContext componentContext) {
         }
         thingTypeRegistryServiceTracker.close();
         configDescriptionRegistryServiceTracker.close();
-        this.thingHandlers.clear();
-        this.configStatusProviders.clear();
-        this.firmwareUpdateHandlers.clear();
-        this.bundleContext = null;
+        configStatusProviders.clear();
+        firmwareUpdateHandlers.clear();
+        bundleContext = null;
     }
 
     @Override
-    public void unregisterHandler(Thing thing) {
-        ServiceRegistration<ThingHandler> thingHandlerServiceRegistration = thingHandlers
-                .remove(thing.getUID().toString());
-        if (thingHandlerServiceRegistration != null) {
-            logger.debug("Removing service registration for handler of thing '{}'.", thing.getUID().getAsString());
-            unregisterHandler(thingHandlerServiceRegistration);
-        } else {
-            logger.error("There is no service registration for a thing handler of thing '{}'.",
-                    thing.getUID().getAsString());
-        }
-        ServiceRegistration<ConfigStatusProvider> configStatusProviderServiceRegistration = configStatusProviders
-                .remove(thing.getUID().getAsString());
-        if (configStatusProviderServiceRegistration != null) {
-            configStatusProviderServiceRegistration.unregister();
-        }
-        ServiceRegistration<FirmwareUpdateHandler> firmwareUpdateHandlerServiceRegistration = firmwareUpdateHandlers
-                .remove(thing.getUID().getAsString());
-        if (firmwareUpdateHandlerServiceRegistration != null) {
-            firmwareUpdateHandlerServiceRegistration.unregister();
-        }
-    }
+    public ThingHandler registerHandler(Thing thing) {
+        Preconditions.checkArgument(thing != null, "The argument 'thing' must not be null.");
 
-    private void unregisterHandler(ServiceRegistration<ThingHandler> serviceRegistration) {
-        ThingHandler thingHandler = bundleContext.getService(serviceRegistration.getReference());
-        removeHandler(thingHandler);
-        serviceRegistration.unregister();
-        if (thingHandler instanceof BaseThingHandler) {
-            ((BaseThingHandler) thingHandler).unsetBundleContext(bundleContext);
-        }
-    }
-
-    @Override
-    public void registerHandler(Thing thing, ThingHandlerCallback thingHandlerListener) {
         ThingHandler thingHandler = createHandler(thing);
         if (thingHandler == null) {
             throw new IllegalStateException(this.getClass().getSimpleName()
                     + " could not create a handler for the thing '" + thing.getUID() + "'.");
         }
+        setHandlerContext(thingHandler);
```

Comment:
Removing the bundle context from the `BaseThingHandler` causes the following refactoring:

- The `BaseThingHandlerFactory` must handle the tracking of the `ThingRegistry` and the `ItemChannelLinkRegistry`.
- Due to the fact that the `BaseThingHandlerFactory` is an abstract class we can not use declarative service features. This would end in a crappy implementation (service tracker must be used, the factory must be  unregistered if the `ThingRegistry` disappears, etc.)

This leads me to the following question: should a binding developer use core services like ThingRegistry, ItemChannelLinkRegistry and ItemRegistry? 

My proposal would be:
- A binding should not use ESH core services directly.
- Required functionality of such core services could be provided by the framework.
- E.g. the `ThingHandlerCallback` provides methods (currently used in the `BaseThingHandler`) like
	- `ThingHandlerCallback.getBridge(ThingUID bridgeUID)`
	- `ThingHandlerCallback.isItemLinked(String channelId)`
	- ... further use cases TBD
- `BundleContext`, `ThingRegistry`, `ItemChannelLinkRegistry` would be removed from the `BaseThingHandler`.

Problem:
- This will break a lot of bindings, because `BundleContext`, `ThingRegistry`, `ItemChannelLinkRegistry` are protected members in the `BaseThingHandler`. 
- `ThingHandlerCallback.getBridge(ThingUID bridgeUID)` and `ThingHandlerCallback.isItemLinked(String channelId)` are the obvious use cases covered by the `BaseThingHandler`. Further uses cases must be analyzed (which bindings are using the protected members for what?).

Due to the fact the this refactoring will break a lot of bindings, I would like to provide this refactoring in a separate PR. The changes of this PR are quite enough. 

@kaikreuzer what do you think?